client/src/nonblocking.rs

```diff
@@ -0,0 +1,253 @@
+    ) -> Result<u8, ClientError> {
+        let addresses = vec![self.program_id.to_string()];
+        let filter = RpcTransactionLogsFilter::Mentions(addresses);
+        let self_program_str = self.program_id.to_string();
+        let config = self.config.clone();
+        let (unsubscribe_sender, unsubscribe_receiver) = unbounded_channel::<_>();
+
+        let handle = spawn({
+            let client = Arc::clone(&self.client);
+
+            async move {
+                let (mut notifications, unsubscribe) =
+                    client.logs_subscribe(filter, config).await?;
+
+                unsubscribe_sender.send(unsubscribe).map_err(|e| {
+                    ClientError::SolanaClientPubsubError(PubsubClientError::UnexpectedMessageError(
+                        e.to_string(),
+                    ))
+                })?;
+
+                while let Some(logs) = notifications.next().await {
+                    let ctx = EventContext {
+                        signature: logs.value.signature.parse().unwrap(),
+                        slot: logs.context.slot,
+                    };
+                    let mut logs = &logs.value.logs[..];
+                    if !logs.is_empty() {
+                        if let Ok(mut execution) = Execution::new(&mut logs) {
+                            for l in logs {
+                                // Parse the log.
+                                let (event, new_program, did_pop) = {
+                                    if self_program_str == execution.program() {
+                                        handle_program_log(&self_program_str, l).unwrap_or_else(
+                                            |e| {
+                                                println!("Unable to parse log: {e}");
+                                                std::process::exit(1);
+                                            },
+                                        )
+                                    } else {
+                                        let (program, did_pop) =
+                                            handle_system_log(&self_program_str, l);
+                                        (None, program, did_pop)
+                                    }
+                                };
+                                // Emit the event.
+                                if let Some(e) = event {
+                                    f(&ctx, e);
+                                }
+                                // Switch program context on CPI.
+                                if let Some(new_program) = new_program {
+                                    execution.push(new_program);
+                                }
+                                // Program returned.
+                                if did_pop {
+                                    execution.pop();
+                                }
+                            }
+                        }
+                    }
```

Comment:
Is this part the same with blocking client? If so can we reuse functionality and not duplicate the code?